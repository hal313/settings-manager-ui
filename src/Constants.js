const ATTRIBUTE_PREFIX = 'data-setting';

const Constants = {
    ATTRIBUTE_PREFIX: ATTRIBUTE_PREFIX,
    ATTRIBUTE_NAME: `${ATTRIBUTE_PREFIX}-name`,
    ATTRIBUTE_TYPE: `${ATTRIBUTE_PREFIX}-type`,
    ATTRIBUTE_CONTAINER_ELEMENT: `${ATTRIBUTE_PREFIX}-container`
};

export {Constants};