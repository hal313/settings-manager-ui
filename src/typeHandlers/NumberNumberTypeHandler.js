import {isDefinedAndNotEmpty} from '../util/isDefinedAndNotEmpty.js';
import {createError} from '../util/createError.js';
import { isElement } from '../util/isElement.js';
import { isNumber } from '../util/isNumber.js';
import { TypeHandler } from './TypeHandler.js';

export class NumberNumberTypeHandler extends TypeHandler
{

    constructor() {
        super();
    }

    getType() {
        return NumberNumberTypeHandler.TYPE;
    }

    getValue(element) {
        return +element.value;
    }

    setValue(element, value) {
        if (!isElement(element)) {
            throw createError('The "element" parameter must be an Element');
        }

        element.value = +value;
    }

    createElement(name, value) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('input');
        element.setAttribute('name', name);
        element.setAttribute('type', 'number');
        element.setAttribute('value', value);

        this.setValue(element, value);

        return element;
    }

};
NumberNumberTypeHandler.TYPE = 'number:number';