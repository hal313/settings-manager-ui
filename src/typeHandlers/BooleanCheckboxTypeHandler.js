import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';

export class BooleanCheckboxTypeHandler {

    constructor() {}

    getType() {
        return 'boolean:checkbox';
    }

    getValue(element) {
        // return !!element.value;
        return !!element.hasAttribute('checked');
    }

    setValue(element, value) {
        // element.value = !!value
        !!value ? element.setAttribute('checked', '') : element.removeAttribute('checked');
    }

    createElement(name, value) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        // return asElement(`<input name="${name}" ${Constants.ATTRIBUTE_NAME}="${name}" ${Constants.ATTRIBUTE_TYPE}="${this.getType()}" type="checkbox" ${!!value?'checked':''}/> ${name}`);
        let element = document.createElement('input');
        element.setAttribute('type', 'checkbox');
        element.setAttribute('name', name);
        if (!!value) {
            element.setAttribute('checked', '');
        }
        element.setAttribute(Constants.ATTRIBUTE_NAME, name);
        element.setAttribute(Constants.ATTRIBUTE_TYPE, this.getType());
        return element;
    }

};