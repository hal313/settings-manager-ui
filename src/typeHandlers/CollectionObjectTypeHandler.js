import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';
import { isObjectArray } from '../util/isObjectArray.js';
import { getChildSettingElements } from '../util/getChildSettingElements.js';
import { getNameFromElement } from '../util/getNameFromElement.js';
import { getOneElement } from '../util/getOneElement.js';
import { getAllElements } from '../util/getAllElements.js';
import { isDefined } from '../util/isDefined.js';
import { SettingModifier } from '../util/SettingModifier.js';

const ATTRIBUTE_INDEX = 'data-setting-index';
const ATTRIBUTE_COLLECTION_VALUE_CONTAINER = 'data-setting-value-container';


/**
 * Gets the index from an element.
 *
 * @param {Element} element the element to get the index from
 * @returns {Number} the index assigned to the element, based on the attribute
 * @throws when the index is not present
 */
const getIndexFrom = (element) => {
    let index = element.getAttribute(ATTRIBUTE_INDEX);
    if (isDefined(index)) {
        return +index;
    }
    throw createError(`element does not have an index assigned via ${ATTRIBUTE_INDEX} "${element.outerHTML}"`);
};

/**
 * Gets an array item container element by index.
 *
 * @param {Element} rootElement the element to get the array item container element from
 * @param {Number} index the index of the array item container element to get
 * @returns {Element} the element marked as having the specified index
 */
const getArrayItemContainerElementByIndex = (rootElement, index) => {
    return getOneElement(`[${ATTRIBUTE_INDEX}="${index}"]`, rootElement);
};

/**
 * Gets an array item container element by index.
 *
 * @param {Element} rootElement the element to get the array item container element from
 * @param {Number} index the index of the array item container element to get
 * @returns {Element[]} the elements marked as having the specified index
 */
const getArrayItemContainersElementByIndex = (rootElement, index) => {
    return getAllElements(`[${ATTRIBUTE_INDEX}="${index}"]`, rootElement);
};

/**
 * Gets an array item container element
 *
 * @param {Element} rootElement the root element
 * @returns {Element[]} all elements which are root containers
 */
const getArrayItemContainerElements = (rootElement) => {
    return getAllElements(`[${ATTRIBUTE_COLLECTION_VALUE_CONTAINER}]`, rootElement);

}

/**
 * Creates an slement which is suitable for residing in a DOM and represents the array element value.
 *
 * @param {*} value the value to create an element for
 * @param {Number} index the index of the item
 * @param {SettingModifier} settingModifier the SettingModifer to assist
 * @returns {Element} the created element
 */
const createArrayItemContainerElement = (value, index, settingModifier) => {
    const name = `name_${index}`;
    const element = settingModifier.createElement(name, value, settingModifier);
    element.setAttribute(ATTRIBUTE_INDEX, index);
    element.setAttribute(ATTRIBUTE_COLLECTION_VALUE_CONTAINER, '');
    return element;
};

export class CollectionObjectTypeHandler {

    constructor() {}

    getType() {
        return 'collection:object';
    }

    getValue(element, settingModifier) {
        let childElements = getChildSettingElements(element);
        let valueArray = [];

        // For each child setting, add the name and value touple to the returned settings object
        childElements.forEach((childElement) => {
            // The value object represented by this child
            const valueObject = {};

            // The element may be
            const valueElement = getChildSettingElements(childElement)[0];
            // Get the name from the element
            const name = getNameFromElement(valueElement);
            // Get the value (which may be a recursive call)
            const value = settingModifier.getValue(valueElement);

            // Add the name and value to the value object
            valueObject[name] = value;

            // Push the object
            valueArray.push(valueObject);
        });

        return valueArray;
    }

    setValue(element, settings, settingModifier) {
        // TODO: Is this necessary?
        element.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');

        // Get the container elements?
        let arrayItemContainerElements = getArrayItemContainerElements(element);

        // Remove values from the DOM which do not exist in the value
        arrayItemContainerElements.forEach((arrayItemContainerElement) => {
            let index = getIndexFrom(arrayItemContainerElement);

            // If the index of the element is more than the length if the settings, remove it
            // because that means the new settings array is shorter than the existing DOM array
            if ((index + 1) > settings.length) {
                arrayItemContainerElement.remove();
            }
        });


        // TODO: Merge dom and settings so they stay in order based on index
        // Go through the settings array; if the nth element is in the dom, set the value; otherwise, append it at the end
        settings.forEach((setting, index) => {
            let arrayItemContainerElements = getArrayItemContainersElementByIndex(element, index);

            switch (arrayItemContainerElements.length) {
                case 0:
                    // Add a new DOM element
                    element.appendChild(createArrayItemContainerElement(setting, index, settingModifier));
                    break;
                case 1:
                    // Update an existing DOM element
                    let arrayItemContainerElement = getArrayItemContainerElementByIndex(element, index);
                    settingModifier.setValue(arrayItemContainerElement, setting);
                    break;
                default:
                    throw createError(`Duplicate elements in setting "${getNameFromElement(element)}"`);
            }
        });

        element.value = settings;
    }

    createElement(name, value, settingModifier) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }
        if (!isObjectArray(value)) {
            throw createError('The "value" parameter must be an array');
        }

        // Create the root element
        const rootElement = document.createElement('div');

        // Set the value
        rootElement.value = value;

        // Set attributes
        rootElement.setAttribute('name', name);
        rootElement.setAttribute(Constants.ATTRIBUTE_NAME, name);
        rootElement.setAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT, '');
        rootElement.setAttribute(Constants.ATTRIBUTE_TYPE, this.getType());

        // Set the value
        this.setValue(rootElement, value, settingModifier);

        return rootElement;
    }

};
