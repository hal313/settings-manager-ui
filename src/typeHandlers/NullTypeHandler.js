import { isDefinedAndNotEmpty } from '../util/isDefinedAndNotEmpty.js';
import { createError } from '../util/createError.js';
import { Constants } from '../Constants.js';

export class NullTypeHandler {

    constructor() {}

    getType() {
        return 'null';
    }

    getValue(/*element*/) {
        return null;
    }

    setValue(element/*, value*/) {
        element.value = null;
    }

    createElement(name/* , value */) {
        if (!isDefinedAndNotEmpty(name)) {
            throw createError('The "name" parameter must be specified');
        }

        let element = document.createElement('div');
        element.setAttribute('name', name);
        element.setAttribute('value', null);

        this.setValue(element, null);

        return element;
    }

};