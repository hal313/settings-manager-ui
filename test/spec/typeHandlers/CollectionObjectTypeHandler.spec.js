import { CollectionObjectTypeHandler } from '../../../src/typeHandlers/CollectionObjectTypeHandler';
import { SettingModifier } from '../../../src/util/SettingModifier';
import { ObjectDefaultTypeHandler } from '../../../src/typeHandlers/ObjectDefaultTypeHandler';
import { TypeHandlerManager } from '../../../src/util/TypeHandlerManager';
import { StringTextTypeHandler } from '../../../src/typeHandlers/StringTextTypeHandler';
import { BooleanCheckboxTypeHandler } from '../../../src/typeHandlers/BooleanCheckboxTypeHandler';
import { NumberNumberTypeHandler } from '../../../src/typeHandlers/NumberNumberTypeHandler';
import { TypeDecoratorManager } from '../../../src/util/TypeDecoratorManager';
import { Constants } from '../../../src/Constants';
import { StringPasswordTypeHandler } from '../../../src/typeHandlers/StringPasswordTypeHandler';
import { UndefinedTypeHandler } from '../../../src/typeHandlers/UndefinedTypeHandler';
import { NullTypeHandler } from '../../../src/typeHandlers/NullTypeHandler';

// TODO: Test types (or cast to boolean)
// TODO: Handle bad cases (null element, etc)
describe('CollectionObjectTypeHandler', () => {

    describe('Lifecycle', () => {

        test('should exist', () => {
            expect(CollectionObjectTypeHandler).toBeDefined();
        });

    });

    describe('API', () => {
        let collectionObjectTypeHandler;
        let typeHandlerManager;
        let typeDecoratorManager;
        let settingModifier;

        beforeEach(() => {
            collectionObjectTypeHandler = new CollectionObjectTypeHandler();

            typeHandlerManager = new TypeHandlerManager();

            const stringTextTypeHandler = new StringTextTypeHandler();
            const booleanCheckboxTypeHandler = new BooleanCheckboxTypeHandler();
            const numberNumberTypeHandler = new NumberNumberTypeHandler();
            const objectDefaultTypeHandler = new ObjectDefaultTypeHandler();
            const undefinedTypeHandler = new UndefinedTypeHandler();
            const nullTypeHandler = new NullTypeHandler();
            const stringPasswordTypeHandler = new StringPasswordTypeHandler();


            typeHandlerManager.addTypeHandler(stringTextTypeHandler);
            typeHandlerManager.addTypeHandler(booleanCheckboxTypeHandler);
            typeHandlerManager.addTypeHandler(numberNumberTypeHandler);
            typeHandlerManager.addTypeHandler(objectDefaultTypeHandler);
            typeHandlerManager.addTypeHandler(undefinedTypeHandler);
            typeHandlerManager.addTypeHandler(nullTypeHandler);
            typeHandlerManager.addTypeHandler(collectionObjectTypeHandler);
            typeHandlerManager.addTypeHandler(stringPasswordTypeHandler);


            typeHandlerManager.setDefaultHandler('string', stringTextTypeHandler.getType());
            typeHandlerManager.setDefaultHandler('boolean', booleanCheckboxTypeHandler.getType());
            typeHandlerManager.setDefaultHandler('number', numberNumberTypeHandler.getType());
            typeHandlerManager.setDefaultHandler('object', objectDefaultTypeHandler.getType());
            typeHandlerManager.setDefaultHandler('collection', collectionObjectTypeHandler.getType());

            typeDecoratorManager = new TypeDecoratorManager();

            settingModifier = new SettingModifier(typeHandlerManager, typeDecoratorManager);
        });

        describe('createElement', () => {

            test('should throw when name is undefined', () => {
                expect(() => collectionObjectTypeHandler.createElement(undefined, 'someValue')).toThrowError();
            });

            test('should throw when name is null', () => {
                expect(() => collectionObjectTypeHandler.createElement(null, 'someValue')).toThrowError();
            });

            test('should throw when name is the empty string', () => {
                expect(() => collectionObjectTypeHandler.createElement('', 'someValue')).toThrowError();
            });

            test('should throw when name is whitespace', () => {
                expect(() => collectionObjectTypeHandler.createElement('   ', 'someValue')).toThrowError();
            });

            test('should create an element which holds array values', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Check the type
                expect(element instanceof Element).toBeTruthy();
                // Check the required attributes
                expect(element.getAttribute(Constants.ATTRIBUTE_NAME)).toEqual(name);
                expect(element.getAttribute(Constants.ATTRIBUTE_CONTAINER_ELEMENT)).toBeDefined();
                // Check the value
                expect(element.value).toEqual(value);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

        describe('getValue', () => {

            test('should get the value from a self-created element', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Check the value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
            });

        });

        describe('setValue', () => {

            test('should set the value for a self-created element', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];
                const newValue = [
                    {first: 11},
                    {second: false},
                    {third: '33'}
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                collectionObjectTypeHandler.setValue(element, newValue, settingModifier);

                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(newValue);

                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test.skip('should set the value for a self-created element with changing types', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];
                const newValue = [
                    {first: 'one'},
                    {second: false},
                    {third: 3}
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();


                // Set the value
                collectionObjectTypeHandler.setValue(element, newValue, settingModifier);


                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Check the value
                expect(element.value).toEqual(newValue);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should set the value for a self-created element when one element is removed', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];
                const newValue = [
                    {first: 1},
                    {second: false}
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                collectionObjectTypeHandler.setValue(element, newValue, settingModifier);

                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(newValue);

                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            // TODO: Test embedded ({firstName: 1, lastName: 1, optional: {middle: 3}})
            test('should set the value for a self-created element when one element is changed', () => {
                const name = 'people';
                const value = [
                    {firstName: 'first name 1', lastName: 'last name 1'},
                    {firstName: 'first name 2', lastName: 'last name 2'},
                    {firstName: 'first name 3', lastName: 'last name 3'},
                ];
                const newValue = [
                    {firstName: 'first name 1', lastName: 'last name 1'},
                    {firstName: 'first name 2', lastName: 'last name 2'},
                    {firstName: 'first name CHANGED', lastName: 'last name CHANGED'},
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                collectionObjectTypeHandler.setValue(element, newValue, settingModifier);

                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(newValue);
                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test('should set the value for a self-created element when one element is added', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];
                const newValue = [
                    {first: 11},
                    {second: false},
                    {third: '33'},
                    {fourth: 'four'}
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier, {});

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier, {})).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                collectionObjectTypeHandler.setValue(element, newValue, settingModifier, {});

                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier, {})).toEqual(newValue);

                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

            test.skip('should set the value for a self-created element when one element is added with changing types', () => {
                const name = 'fieldName';
                const value = [
                    {first: 1},
                    {second: true},
                    {third: '3'}
                ];
                const newValue = [
                    {first: 'one'},
                    {second: false},
                    {third: 3},
                    {fourth: 'four'}
                ]

                // Create the element
                const element = collectionObjectTypeHandler.createElement(name, value, settingModifier);

                // Precondition check
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(value);
                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();

                collectionObjectTypeHandler.setValue(element, newValue, settingModifier);

                // Check the new value
                expect(collectionObjectTypeHandler.getValue(element, settingModifier)).toEqual(newValue);

                // Check the value
                expect(element.value).toEqual(newValue);

                // Match the HTML
                expect(element.outerHTML).toMatchSnapshot();
            });

        });

    });

});