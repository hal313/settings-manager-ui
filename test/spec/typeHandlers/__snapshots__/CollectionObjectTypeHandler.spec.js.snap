// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectionObjectTypeHandler API createElement should create an element which holds array values 1`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="true"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element 1`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="true"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element 2`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="false"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is added 1`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="true"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is added 2`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="false"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
  <div name="fieldName[3]"
       data-setting-container
       class
       data-setting-flatname="fieldName[3]"
       data-setting-name="fieldName[3]"
       data-setting-type="object:default"
       data-setting-index="3"
       data-setting-value-container
  >
    <input name="fourth"
           type="text"
           value="four"
           data-setting-flatname="fieldName[3].fourth"
           data-setting-name="fourth"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is changed 1`] = `

<div name="people"
     data-setting-container
     data-setting-name="people"
     class
>
  <div name="people[0]"
       data-setting-container
       class
       data-setting-flatname="people[0]"
       data-setting-name="people[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 1"
           data-setting-flatname="people[0].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 1"
           data-setting-flatname="people[0].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
  <div name="people[1]"
       data-setting-container
       class
       data-setting-flatname="people[1]"
       data-setting-name="people[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 2"
           data-setting-flatname="people[1].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 2"
           data-setting-flatname="people[1].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
  <div name="people[2]"
       data-setting-container
       class
       data-setting-flatname="people[2]"
       data-setting-name="people[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 3"
           data-setting-flatname="people[2].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 3"
           data-setting-flatname="people[2].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is changed 2`] = `

<div name="people"
     data-setting-container
     data-setting-name="people"
     class
>
  <div name="people[0]"
       data-setting-container
       class
       data-setting-flatname="people[0]"
       data-setting-name="people[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 1"
           data-setting-flatname="people[0].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 1"
           data-setting-flatname="people[0].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
  <div name="people[1]"
       data-setting-container
       class
       data-setting-flatname="people[1]"
       data-setting-name="people[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 2"
           data-setting-flatname="people[1].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 2"
           data-setting-flatname="people[1].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
  <div name="people[2]"
       data-setting-container
       class
       data-setting-flatname="people[2]"
       data-setting-name="people[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="firstName"
           type="text"
           value="first name 3"
           data-setting-flatname="people[2].firstName"
           data-setting-name="firstName"
           data-setting-type="string:text"
    >
    <input name="lastName"
           type="text"
           value="last name 3"
           data-setting-flatname="people[2].lastName"
           data-setting-name="lastName"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is removed 1`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="true"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
  <div name="fieldName[2]"
       data-setting-container
       class
       data-setting-flatname="fieldName[2]"
       data-setting-name="fieldName[2]"
       data-setting-type="object:default"
       data-setting-index="2"
       data-setting-value-container
  >
    <input name="third"
           type="text"
           value="3"
           data-setting-flatname="fieldName[2].third"
           data-setting-name="third"
           data-setting-type="string:text"
    >
  </div>
</div>

`;

exports[`CollectionObjectTypeHandler API setValue should set the value for a self-created element when one element is removed 2`] = `

<div name="fieldName"
     data-setting-container
     data-setting-name="fieldName"
     class
>
  <div name="fieldName[0]"
       data-setting-container
       class
       data-setting-flatname="fieldName[0]"
       data-setting-name="fieldName[0]"
       data-setting-type="object:default"
       data-setting-index="0"
       data-setting-value-container
  >
    <input name="first"
           type="number"
           value="1"
           data-setting-flatname="fieldName[0].first"
           data-setting-name="first"
           data-setting-type="number:number"
    >
  </div>
  <div name="fieldName[1]"
       data-setting-container
       class
       data-setting-flatname="fieldName[1]"
       data-setting-name="fieldName[1]"
       data-setting-type="object:default"
       data-setting-index="1"
       data-setting-value-container
  >
    <input type="checkbox"
           name="second"
           value="false"
           data-setting-flatname="fieldName[1].second"
           data-setting-name="second"
           data-setting-type="boolean:checkbox"
    >
  </div>
</div>

`;
